libp2p/crypto/rsa.nim

```diff
@@ -560,29 +573,32 @@ proc init*[T: RsaPKI](sospk: var T, data: string): Result[void, Asn1Error] {.inl
   ## Procedure returns ``Result[void, Asn1Status]``.
   sospk.init(fromHex(data))
 
-proc init*(t: typedesc[RsaPrivateKey], data: openarray[byte]): RsaPrivateKey =
+proc init*(t: typedesc[RsaPrivateKey], data: openarray[byte]): RsaResult[RsaPrivateKey] =
   ## Initialize RSA private key from ASN.1 DER binary representation ``data``
   ## and return constructed object.
-  let res = result.init(data)
-  if res.isErr:
-    raise newException(RsaKeyIncorrectError,
-                       "Incorrect private key (" & $res.error & ")")
+  var res: RsaPrivateKey
+  if res.init(data).isErr:
+    err(RsaKeyIncorrectError)
+  else:
+    ok(res)
 
-proc init*(t: typedesc[RsaPublicKey], data: openarray[byte]): RsaPublicKey =
+proc init*(t: typedesc[RsaPublicKey], data: openarray[byte]): RsaResult[RsaPublicKey] =
   ## Initialize RSA public key from ASN.1 DER binary representation ``data``
   ## and return constructed object.
-  let res = result.init(data)
-  if res.isErr:
-    raise newException(RsaKeyIncorrectError,
-                       "Incorrect public key (" & $res.error & ")")
+  var res: RsaPublicKey
+  if res.init(data).isErr:
```

Comment:
The original error in here we are checking actually was `Asn1Error`
Not sure what to do to propagate multiple errors, this kind fits the theory of using `string` so to have a common error format
@arnetheduck

Reply:
see `mapErrTo` - ie when using enum's, the custom is to translate the error to a "local" representation which often aggregates what happened in the layer below so that it makes sense for the current layer and whoever is calling it, in terms of granularity for example - what the error was in the layer below is an implementation detail - if this API needs such detail in its error reporting, it should "recreate" errors in its local enum - this is a bit of a chore and there are ways around that (I'm sure we can come up with some macro should we ever need it - @alehander42 was in fact experimenting with one), but it's good enough as a starting point.

the alternative is indeed a string where the error remains a diagnostic _only_ - but I wouldn't sweat it in this case.

another thing is that for these `init` functions, in bls I renamed them to `fromDer`, `fromRaw` etc since it's slightly ambiguous what format `data` is in, so might as well be explicit.

Reply:
Tried using those helpers, in this case all of them made it look worse

Reply:
yeah, I've deliberately not made them more fancy yet, want to establish a need first so some "ugliness" can be expected